src/tree/binary_tree/mod.rs

```diff
@@ -0,0 +1,310 @@
+        let mut d = vec![node_index];
+        while x != r {
+            x = self.parent(x)?;
+            d.push(x);
+        }
+        Ok(d)
+    }
+
+    /// Copath of a leaf. Ordered from leaf to root. Throws an error if the
+    /// given leaf index is out of bounds.
+    pub(crate) fn copath(&self, leaf_index: LeafIndex) -> Result<Vec<NodeIndex>, BinaryTreeError> {
+        let node_index = NodeIndex::from(leaf_index);
+        // If the tree only has one leaf
+        if node_index == self.root() {
+            return Ok(vec![]);
+        }
+        // Add leaf node
+        let mut d = vec![node_index];
+        // Add direct path
+        d.append(&mut self.leaf_direct_path(leaf_index)?);
+        // Remove root node
+        d.pop();
+        // Calculate copath
+        d.iter()
+            .map(|&node_index| self.sibling(node_index))
+            .collect()
+    }
+
+    /// Get the index of the common ancestor of the nodes corresponding to the
+    /// two given indices. Throws an error if one of the indices is out of
+    /// bounds.
+    pub(crate) fn common_ancestor(
+        &self,
+        x: NodeIndex,
+        y: NodeIndex,
+    ) -> Result<NodeIndex, BinaryTreeError> {
+        self.is_out_of_bounds(x)?;
+        self.is_out_of_bounds(y)?;
+        let (lx, ly) = (level(x) + 1, level(y) + 1);
+        if (lx <= ly) && (x.as_usize() >> ly == y.as_usize() >> ly) {
+            return Ok(y);
+        } else if (ly <= lx) && (x.as_usize() >> lx == y.as_usize() >> lx) {
+            return Ok(x);
+        }
+
+        let (mut xn, mut yn) = (x.as_usize(), y.as_usize());
+        let mut k = 0;
+        while xn != yn {
+            xn >>= 1;
+            yn >>= 1;
+            k += 1;
+        }
+        Ok(NodeIndex::from((xn << k) + (1 << (k - 1)) - 1))
+    }
+
+    /// Replace the node at index `index`, consuming the new node and returning
+    /// the old one. Throws an error if the given index is out of bounds.
+    pub(crate) fn replace(&mut self, node_index: NodeIndex, node: T) -> Result<T, BinaryTreeError> {
+        self.is_out_of_bounds(node_index)?;
+        // First push the node to the end of the nodes array.
+        self.nodes.push(node);
+        // Then use `swap_remove`, which replaces the target node with the one
+        // at the end of the vector.
+        Ok(self.nodes.swap_remove(node_index.as_usize()))
+    }
+
+    /// Add nodes to the tree on the right side such that the tree is still
+    /// left-balanced. The number of nodes added has to be even, as we want the
+    /// tree to remain full. Throws an error if the number of nodes added would
+    /// cause the tree to become non-full, i.e. if the number of nodes added
+    /// modulo 2 is not zero.
+    pub(crate) fn add(&mut self, nodes: Vec<T>) -> Result<(), BinaryTreeError> {
+        if nodes.len() % 2 != 0 {
+            return Err(BinaryTreeError::TreeNotFull);
+        }
+        self.nodes.extend(nodes);
+        Ok(())
+    }
```

Comment:
You could enforce this through type safety as well, e.g. like that:

`fn add(&mut self, nodes: Vec<(T, T)>)`

Reply:
In theory, the tree could be empty (which would be valid, I think, although we could forbid it) and a tree with one node is also a valid tree.

Reply:
Never mind. Allowing an empty tree is not a good idea.